libs/algorithms/include/hpx/parallel/algorithms/for_each.hpp

```diff
@@ -8,23 +8,272 @@
 
 #pragma once
 
+#if defined(DOXYGEN)
+namespace hpx {
+    /// Applies \a f to the result of dereferencing every iterator in the
+    /// range [first, last).
+    ///
+    /// \note   Complexity: Applies \a f exactly \a last - \a first times.
+    ///
+    /// If \a f returns a result, the result is ignored.
+    ///
+    /// If the type of \a first satisfies the requirements of a mutable
+    /// iterator, \a f may apply non-constant functions through the
+    /// dereferenced iterator.
+    ///
+    /// \tparam InIter      The type of the source begin and end iterator used
+    ///                     (deduced). This iterator type must meet the
+    ///                     requirements of an input iterator.
+    /// \tparam F           The type of the function/function object to use
+    ///                     (deduced). F must meet requirements of
+    ///                     \a MoveConstructible.
+    /// \tparam Proj        The type of an optional projection function. This
+    ///                     defaults to \a util::projection_identity
+    /// \param first        Refers to the beginning of the sequence of elements
+    ///                     the algorithm will be applied to.
+    /// \param last         Refers to the end of the sequence of elements the
+    ///                     algorithm will be applied to.
+    /// \param f            Specifies the function (or function object) which
+    ///                     will be invoked for each of the elements in the
+    ///                     sequence specified by [first, last).
+    ///                     The signature of this predicate
+    ///                     should be equivalent to:
+    ///                     \code
+    ///                     <ignored> pred(const Type &a);
+    ///                     \endcode \n
+    ///                     The signature does not need to have const&. The
+    ///                     type \a Type must be such that an object of
+    ///                     type \a FwdIter can be dereferenced and then
+    ///                     implicitly converted to Type.
+    /// \param proj         Specifies the function (or function object) which
+    ///                     will be invoked for each of the elements as a
+    ///                     projection operation before the actual predicate
+    ///                     \a f is invoked.
+    ///
+    /// \returns            \a f.
+    template <typename InIter, typename F,
+        typename Proj = util::projection_identity>
+    typename F for_each(InIter first, InIter last, F&& f, Proj&& proj = Proj())
```

Comment:
It's just doxygen, but the leading `typename` is superfluous.

Reply:
Also, the base algorithm shouldn't have a projection.

Reply:
Good catch. That's a leftover from copying and pasting.

About the projection, I saw that it's not there in the standard, but should we explicitly leave it out? Are we not following the standard by having that as an extension (it's an optional parameter after all)?

Reply:
I think we should follow the standard by the letter, I think. If somebody needs projections, they can use `hpx::ranges::for_each`.

Reply:
I've removed the projection parameter from these overloads. I've left the non-ranges tests that use projections to use the deprecated `hpx::parallel::for_each`.